File: new-resource-detector/src/main/java/nl/haploid/resource/detector/service/ResourceDetectorsService.java
package nl.haploid.resource.detector.service;

import nl.haploid.event.AtomChangeEvent;
import nl.haploid.resource.detector.detector.ResourceDetector;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ResourceDetectorsService {

	@Autowired
	private List<ResourceDetector> detectors;

	protected List<ResourceDetector> getDetectors() {
		return detectors;
	}

	public List<ResourceDetector> getDetectorsForAtomType(final String atomType) {
		return this.getDetectors().stream()
				.filter(d -> d.getAtomTypes().contains(atomType))
				.collect(Collectors.toList());
	}

	public List<ResourceDescriptor> detectResources(final String atomType, final List<AtomChangeEvent> events) {
		return getDetectorsForAtomType(atomType).stream()
				.map(d -> d.detect(events))
				.flatMap(Collection::stream)
				.collect(Collectors.toList());
	}
}

File: new-resource-detector/src/main/scala/nl/haploid/octowight/channel/scout/service/ResourceDetectorsService.scala
package nl.haploid.octowight.channel.scout.service

import java.util

import nl.haploid.octowight.AtomGroup
import nl.haploid.octowight.channel.scout.detector.ResourceDetector
import nl.haploid.octowight.registry.data.ResourceRoot
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.collection.JavaConverters._

@Service
class ResourceDetectorsService {
  @Autowired private[this] val _detectors: util.List[ResourceDetector] = null

  def detectors: Set[ResourceDetector] = _detectors.asScala.toSet

  def detectorsForAtomCategory(atomCategory: String): Set[ResourceDetector] =
    detectors.filter(_.atomCategories.contains(atomCategory))

  def detectResources(atomGroup: AtomGroup, atomIds: Set[Long]): Set[ResourceRoot] =
    detectorsForAtomCategory(atomGroup.category).flatMap(_.detect(atomGroup, atomIds))
}
